src/EditorSidebar.tsx

```diff
@@ -1,82 +1,64 @@
 import JSONSchemaForm from "./ui/JSONSchemaForm.tsx";
-import { FormProvider, useForm } from "react-hook-form";
+import { FormProvider, useFieldArray, useForm } from "react-hook-form";
 import type { PageComponentData } from "../types.ts";
+import AddNewComponent from "./AddNewComponent.tsx";
 
-function AddNewComponent({ onAddComponent }) {
-  const { componentSchemas } = useEditor();
-  const selectedComponent = useRef("");
+const COMPONENTS_KEY_NAME: "components" = "components" as const;
 
-  return (
-    <div class="py-1 mb-2">
-      Selecione componente para adicionar
-      <div class="flex">
-        <select
-          class={tw`border hover:border-black py-1 px-2 mr-1 rounded transition-colors ease-in`}
-          value={"-1"}
-          onInput={(e) => {
-            selectedComponent.current = e.target.value;
-          }}
-        >
-          {Object.keys(componentSchemas).map((componentName) => (
-            <option value={componentName}>
-              {componentSchemas[componentName]?.title ?? componentName}
-            </option>
-          ))}
-        </select>
-        <Button
-          onClick={() => {
-            onAddComponent(selectedComponent.current);
-            selectedComponent.current = "";
-          }}
-        >
-          Adicionar
-        </Button>
-      </div>
-    </div>
-  );
-}
-
-function mapComponentsToFormData(components: PageComponentData[]) {
-  return components.reduce((curr, component, index) => {
-    curr[index] = component.props;
-    return curr;
-  }, {});
+function mapComponentsToFormData(
+  components: PageComponentData[],
+): ComponentProp[] {
+  return components.map((component) => {
+    return component.props || {};
+  });
 }
 
 function mapFormDataToComponents(
-  formData: FormValues,
+  formData: ComponentProp[],
   currentComponents: PageComponentData[],
 ) {
   const components: PageComponentData[] = [];
   currentComponents.forEach(({ component }, index) => {
     const props = formData[index];
+
     components.push({
       component,
-      props,
+      // Required check this, because the form doesn't handle undefined values
+      props: Object.keys(props).length === 0 ? undefined : props,
     });
   });
 
   return components;
 }
 
-type FormValues = Record<number, Record<string, unknown>>;
+type ComponentProp = Record<string, any>;
+type FormValues = {
+  [COMPONENTS_KEY_NAME]: ComponentProp[];
+};
 
 export default function EditorSidebar() {
   const { componentSchemas, template, components: initialComponents } =
     useEditor();
   const componentsRef = useRef(initialComponents);
+  const components = componentsRef.current;
+
   const methods = useForm<FormValues>({
-    defaultValues: mapComponentsToFormData(initialComponents),
+    defaultValues: {
+      [COMPONENTS_KEY_NAME]: mapComponentsToFormData(initialComponents),
+    },
+  });
+  const { append, fields, remove, swap } = useFieldArray({
+    control: methods.control,
+    name: COMPONENTS_KEY_NAME,
```

Comment:
What do you think about moving most of this logic to a custom hook?

File might be getting too large

Reply:
Take a look at the hook, i didn't like it. The component is creating new objects unnecessarily multiple times, wdyt?